doc/developer_guide.md

```diff
@@ -22,6 +22,30 @@ Then make sure your PATH includes the `~/.local/bin` directory, e.g. with:
 
     export PATH="$PATH":~/.local/bin
 
+### Pre-commit hooks
+
+GitHub pre-commit hooks are scripts that can be triggered automatically by
+Git when certain events occur, such as committing changes to a repository.
+They are typically used to enforce coding standards, perform automated tests,
+or perform other checks on code changes before they are committed.
+
+There are two main components in the setup 1) .pre-commit script and
+2) .github-workflow-script. The github-workflow-script extracts the steps
+directly from the github-workflow,while the .pre-commit script will run
+the extracted commands.
+
+To leverage the setup make both the files executable using the following
+commands in your terminal. Make sure you are in the root directory of the
+project.
+
+    chmod +x .pre-commit
+    chmod +x .github-workflow-script
+
+Now, every time you run git commit, the pre-commit hook script will be
+executed before the commit is created. If any of the pre-commit checks
+fail, the commit will be aborted and you will see an error message
+explaining the issue.
```

Comment:
okay I taught we were not working on this right now like linking with the git hook. I did not do that in my code. Will that be a problem? I will replace the make exec commands with this one.
another thing is that the git does not show any pre-commit file on its own. It has the pre-commit.sample 
would we need to like specify the user to do so or our commit will save and git will remember the change?

Reply:
We don't need any code for that. Just add something like this to the docs:
```markdown
To activate the pre-commit hook symlink the `.pre-commit` file into `.git/hooks/pre-commit`.
E.g. with this command:

    ln -s ../../.pre-commit .git/hooks/pre-commit
```
We cannot commit this symlink anyway, because it goes into Git's data for the repository. So developers must create the link themselves. It should also be left up to developers to decide if they want to run anything on commit.

Yes, git doesn't have any `pre-commit` script by default, on purpose. It's up to the user to install what they need. The `pre-commit.sample` is exactly what it says - an example of such a hook with some commonly-useful checks (which we took into our `.pre-commit` script, BTW).